lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java

```diff
@@ -6,159 +6,237 @@
-      // try harder to find the library
-    }
-    String os = System.getProperty("os.name");
-    String name;
-    if (os.startsWith("Mac")) {
-      name = "libenso_parser.dylib";
-    } else if (os.startsWith("Windows")) {
-      name = "enso_parser.dll";
-    } else {
-      name = "libenso_parser.so";
-    }
+public final class Parser {
+  private Parser() {}
+
+  public static String getWarningMessage(Warning warning) {
+    return getWarningTemplate(warning.getId());
+  }
+
+  public static long isIdentOrOperator(CharSequence input) {
+    return getWorker().isIdentOrOperator(input);
+  }
+
+  public static ByteBuffer parseInputLazy(CharSequence input) {
+    return getWorker().parseInputLazy(input);
+  }
 
-    var whereAmI = Parser.class.getProtectionDomain().getCodeSource().getLocation();
-    File root;
-    try {
-      root = new File(whereAmI.toURI()).getParentFile();
-    } catch (URISyntaxException ex) {
-      root = new File(".").getAbsoluteFile();
+  public static Tree parse(CharSequence input) {
+    return getWorker().parse(input);
+  }
+
+  public static UUID getUuid(long metadata, long nodeOffset, long nodeLength) {
+    long high = getUuidHigh(metadata, nodeOffset, nodeLength);
+    long low = getUuidLow(metadata, nodeOffset, nodeLength);
+    if (high == 0 && low == 0) {
+      // The native interface uses the Nil UUID value as a marker to indicate that no UUID was
+      // attached.
+      // The Nil UUID will never collide with a real UUID generated by any scheme.
+      return null;
     }
-    try {
-      var d = root;
-      File path = null;
-      while (d != null) {
-        path = new File(new File(d, "component"), name);
-        if (path.exists()) break;
-        d = d.getParentFile();
-      }
-      if (d == null || path == null) {
-        throw new LinkageError("Cannot find parser in " + root);
+    return new UUID(high, low);
+  }
+
+  /* Worker-thread state */
+
+  private static final FinalizationManager finalizationManager = new FinalizationManager();
+
+  private static Worker getWorker() {
+    var worker = threadWorker.get();
+    if (worker == null) {
+      finalizationManager.runPendingFinalizers();
+      worker = Worker.create();
+      if (!ImageInfo.inImageBuildtimeCode()) {
+        // At build-time, we eagerly free parser buffers; runtime should start out with an empty
+        // `finalizationManager`.
+        finalizationManager.attachFinalizer(worker, worker.finalizer());
+        threadWorker.set(worker);
       }
-      System.load(path.getAbsolutePath());
-    } catch (NullPointerException | IllegalArgumentException | LinkageError e) {
-      if (searchFromDirToTop(e, root, "target", "rust", "parser-jni", name)) {
+    }
+    return worker;
+  }
+
+  private static final ThreadLocal<Worker> threadWorker = new ThreadLocal<>();
+
+  public static void freeAll() {
+    for (var finalizer : finalizationManager.getRegisteredFinalizers()) finalizer.run();
```

Comment:
I'd move this code into `FinalizationManager.runAllFinalizers()` to keep more of encapsulation.

Reply:
This is intentionally using a low-level API:
`freeAll` is designed to be safe to call at any time. It drops current retained buffers, without clearing the registry. This way, we can call it from `EnsoContext::shutdown` without relying too much on the expectation that there will be only one `EnsoContext` in the process and nothing will use the parser after it shuts down. If that assumption became incorrect in the future, dropping buffers unnecessarily would have some performance impact, but wouldn't result in any state corruption or memory leak.
To support this `getRegisteredFinalizers` exposes the finalizers, allowing the caller here to make use of a property that it knows holds of its finalizers: Calling them more than once would be harmless. `FinalizationManager` could offer a `runAllFinalizersWithoutDeregistering`, but I'd prefer it to provide one low-level method that allows the caller to make use of any known special properties of the finalizers.

Reply:
The fact that finalizers can be called multiple time isn't easily to spot from the code...